### PR TITLE
Add support for uploading multiple files

### PR DESCRIPTION
**Pull request by Syfaro**
This allows for uploading multiple files for media groups or thumbnails. It also fixes a longstanding bug with using URLs. 

There's a few breaking changes:
- `UploadFile` has been renamed to `UploadFiles`, and now takes a `[]RequestFile` instead of `fieldname` and `file`
- All `New*Upload` and `New*Share` have been condensed into `New*`
- All `APIResponse` and errors are returned as pointers

I'd love to get feedback on the API changes or what could be done to further improve things.

### CODE REVIEW

**Comment by HirbodBehnam** [comment]
Hello
Is using `panic` in the upload goroutine a good practice? Doesn't panics crash the whole program unless `recover()` is used?
If you don't want to use panic I have 3 other options you can use:
1. Use panics with [recover](https://blog.golang.org/defer-panic-and-recover). But also you need to change the function to `func (bot *BotAPI) UploadFiles(endpoint string, params Params, files []RequestFile) (apiRes *APIResponse, err error)`. In the recover, set `err = errors.New(fmt.Sprintf("%v", r))`; The function will return with an error. Note that you can't return values from a "defered" function. [Here](https://medium.com/swlh/simple-guide-to-panic-handling-and-recovery-in-golang-72d6181ae3e8) is an code sample. 
2. Use an external error variable. For example before starting the goroutine, define a variable internalError (`var internalError error`). Then in the goroutine, if an error occurs, use `internalError` to store the error and then use `return` to stop the goroutine. Check the error after the line `resp, err := bot.Client.Do(req)`.
3. Use [PipeWriter.CloseWithError](https://golang.org/pkg/io/#PipeWriter.CloseWithError]; I'm not sure if this one is a good practice or not, but you can use `w.CloseWithError(err)` and then simply return from the goroutine. The error will be caught at `resp, err := bot.Client.Do(req)`

**Comment by Syfaro** [comment]
Whoops, I forgot to mention that the panics needed to be removed still.

I wasn't aware of `CloseWithError`, that seems like a perfect solution, thank you!

**Comment by abserari** [comment]
waiting for that

File: bot.go [hunk 1]
```diff
@@ -96,7 +95,7 @@ func buildParams(in Params) (out url.Values) {
 }
 
 // MakeRequest makes a request to a specific endpoint with our token.
-func (bot *BotAPI) MakeRequest(endpoint string, params Params) (APIResponse, error) {
+func (bot *BotAPI) MakeRequest(endpoint string, params Params) (*APIResponse, error) {
```
Comment (tjhorner) [inline bot.go:98]: This smells like a breaking change to me. People running off of master or develop are going to have a bad time.

**Review by tjhorner — APPROVED** [review]
This looks good to me, but what are we going to do about the breaking changes, since the versioning is really weird at the moment?

File: bot.go (hunk 1 above)
Comment (Syfaro) [reply bot.go:98]: Yeah. Not really sure what else to do though, I want to keep consistency through everything and there's already a ton of breaking changes elsewhere so might as well do it now.

**Pull request merged**